Add detailed railway-rail-* types

pastk opened the pull request:
Divided `railway=rail` into several types as per #2844.

The old `railway-rail` style and translation are left for compatibility with old mwms.

@vng Does some railway routing code needs to be updated to work with the new types? AFAIR public transport uses `railway-subway` and `railway-light_rail` only.

Newly generated maps will have new underlying `railway-rail-*` types, but looks/visibility/translations are not changed (i.e. they're identical to the old `railway-rail`), i.e. there is no user-visible change. That's because new styles (looks and visibility) require extensive testing on a wide variety of maps and it's too time-consuming to regenerate a lot of maps locally..
I propose to merge the PR before the next release, so that next release cycle we'll have maps with the new types, it'll ease developing new styles a lot. New translations will be done later in a separate PR too.

I generated and tested on a few smaller maps to ensure that:
- railways looks and translations didn't change for old mwms
- same for new mwms, but underlying classificator types has changed

Also cleaned up some obsolete (or bogus?) mappings (`railway=spur/siding/yard` - no usages according to taginfo).
Removed `railway=razed` as there is no point to show a railway was there if there no traces left, its history.

pastk commented:
This new OSM tags to classificator mapping is quite complicated...
Here are some insights I got working on it (related to #4317).

Tags-matching syntax needs to be expanded. The most important missing operators are `!=` and `|`, e.g. `[usage!=main|branch]`.

`replaced_tags.txt` helps to workaround missing `|` atm, but makes it harder to comprehend the matching ideadby having to look into another file.

When `mapcss-mapping.csv` matching capabilities are improved there will be no good reason to keep `replaced_tags.txt` IMHO (or prove me wrong!).

Styles (kothic) "tags-matching". E.g. `[railway=rail][usage=main][!highspeed][!service]` doesn't work to match `railway|rail|main;[railway=rail][usage=main][!highspeed][!service]`. Its likely a bug of handling `!` in kothic. Still, it shows that this kind of "tag-matching" is conceptually wrong. `mapcss-mapping.csv` contains matching expressions to match OSM tags. OK. Kothic doesn't match OSM tags. It matches matching expressions from `mapcss-mapping.csv` using its own matching expressions :) Because there are no more OSM tags past the generation stage and attempts to pretend they're there lead to confusion and unobvious hard-to-debug behavior. So its better to use classificator types names like `railway-rail-main` in the style files (like we do e.g. in `types_strings.txt`).

Comments in `mapcss-mapping.csv` help to understand tags-matching idea/context.

ATM the lines in `mapcss-mapping.csv` are required to be in type IDs order. But it'll be very good to be able to group the types into logical blocks regardless of their IDs. Still, to help choose lower deprecated IDs for new types one can look into e.g. `data/types.txt`.

Also, I tried to generate my test 18.7Mb mwm (~1500 `railway-rail-*` features) with the new types both in the end of `mapcss-mapping.csv` (new "long" IDs) and reusing "short" deprecated IDs (as in this PR) - the size difference was just 440 bytes..

pastk commented:
- Amended the mapping a bit (as a separate commit for easier reviewing).

- Amended and expanded the unit tests - added there all the matching cases I can think of.

- Added a new non-drawable `railway-motor_vehicle` type instead of `railway-rail-motor_vehicle`. Atm a rail is matched either to `railway-rail` or its subtype `tunnel/bridge/motor_vehicle`. I.e. the `-motor_vehicle` subtype is used both for rendering and as a special routing case. As we'll have many `railway-rail-main/branch/etc` types now it doesn't make sense to duplicate `-motor_vehicle` to all of them, so it's better to have it as a special non-drawable type like `hwtag=*` or `route=shuttle_train` (the implementation of the new `railway-motor_vehicle` is very similar). But the old renderable `railway-rail-motor_vehicle` should be retained for the old maps to work.

pastk commented:
> This new OSM tags to classificator mapping is quite complicated... Here are some insights I got working on it (related to #4317).
> 

Another insight - it'll be very good for the maps generation process to produce types combinations stats, similar to `generator_tool -dump_types` output, but for the whole world instead of single mwm. Its useful to discover more rare mapping cases and take them into account, also useful to detect sometimes unwanted cases of two+ types matched to a single feature.

Even better would be to have stats showing all matched types combinations together with original OSM tags combinations that were used for matching.

vng commented on data/mapcss-mapping.csv:
```diff
@@ -96,7 +101,8 @@ leisure|pitch;69;
 highway|motorway|bridge;[highway=motorway][bridge?];;name;int_name;72;
-deprecated|deprecated;73;x
+# -service includes all service=siding/yard/crossover except =spur (see replaced_tags.txt).
+railway|rail|service;[railway=rail][service=service];;name;int_name;73;
```
UPD: Put a comment here that service=service goes from replace_tags.txt
And it has more than 1M entries in total.

pastk replied:
but there is a comment already?

vng replied:
Ok :)

vng commented on data/mapcss-mapping.csv:
```diff
@@ -375,14 +394,17 @@ piste:type|downhill|advanced;[piste:type=downhill][piste:difficulty=advanced];;n
-deprecated|deprecated;351;x
+# -branch includes ordinary railway=rail w/o any specific tags like usage= or service=.
+# Using service= together with usage=branch is a mismapping, but people do it, so exclude explicitly.
+railway|rail|branch;[railway=rail][usage=branch][!highspeed][!service],[railway=rail][!usage][!service][!highspeed];;name;int_name;351;
```
So put exact this line into 28, former railway=rail.
And branch=bridge/tunnel into former rail=bridge/tunnel.

pastk replied:
but then `railway=rail usage/service=rare_unsupported_value` combinations will not be matched by any line...
and we need to show users a railway line there

please take a look into matching cases I put into unit tests..

vng replied:
BTW, also you can delete old railway=rail from styles with this change.

vng commented on data/mapcss-mapping.csv:
```diff
@@ -37,7 +54,7 @@ power|tower;10;
 landuse|grass;12;
 highway|secondary;13;
-natural|wood;14;landuse|forest
+deprecated:natural|wood;14;landuse|forest
```
Don't know why you added "deprecated:" here, but actually, this entry defines landuse|forest.

pastk replied:
It doesn't define anything, its a deprecated type. The [natural=wood] is matched directly in the `landuse|forest` line just above.
I added `deprecated:` so that its actually visible that its deprecated and could be re-used for something else..

vng replied:
This entry should **not** be reused.

pastk replied:
never? why? (I can add a comment with an explanation)

vng replied:
because it is already landuse-forest. and 14 index is used for it

pastk replied:
mmm.. it was deprecated and replaced with `landuse-forest`
this means that the generator ignores this line completely
but the app, when it loads data/types.txt, replaces all id=14 in old mwms to id=11 (of the `landuse-forest`)
so this line `natural|wood;14;landuse|forest` is needed only to continue showing forest tagged with natural=wood for old mwms that were generated before the deprecation change
the change was done Jan 2020, there were no Organic Maps yet :) so its now safe to be re-used
what's wrong in this logic?

vng replied:
Yes, landuse|forest is not a good example.

vng commented on generator/generator_tests/osm_type_test.cpp:
```diff
@@ -473,6 +473,21 @@ UNIT_CLASS_TEST(TestWithClassificator, OsmType_Layer)
+    TEST(params.IsTypeExist(GetType({"highway", "secondary", "bridge"})), ());
+    TEST_EQUAL(params.layer, 1, ());
+  }
+  */
```
I'd prefer to trust OSM here. There are a lot of multi-level junctions, so it's not safe just to assign 1.

pastk replied:
not sure what you mean...
atm we assign `layer=1` automatically for `bridge=yes` (see the test just above)
why not assign it to e.g. `bridge=viaduct` also?

vng replied:
hm, I missed that

vng commented on data/mapcss-mapping.csv:
```diff
@@ -1062,8 +1062,12 @@ highway|speed_camera;991;
+# (because we don't want to have two drawable types in one feature,
+# e.g. railway-rail-motor_vehicle and railway-rail-main|branch|etc)
+# but should be retained for some time to be drawable in old mwms.
+railway|rail|motor_vehicle;[railway=rail][dont=match];;name;int_name;995;
```
It can (should) be replaced with ;railway|rail|main

vng replied:
Also routing code should be updated: RailwayRailMotorVehicle = 994
BTW, it has an error now ...

pastk replied:
so you mean abolish the `railway-rail-motor_vehicle` type and associated special routing code?

vng replied:
This railway will be drawn like main or branch. and you can delete it from styles.
Routing will be broken for old MWMs anyway, but this is not a big problem.

vng replied:
routing logic should be modified on new type

pastk replied:
I believe I did it already in https://github.com/organicmaps/organicmaps/pull/5390/commits/8dbf61b94eb82fde08e347cbdc0c36ecf3967471

vng replied:
Put here:
railway|rail|motor_vehicle;995;railway|rail

pastk commented on data/mapcss-mapping.csv:
```diff
@@ -269,7 +269,8 @@ sport|basketball;242;
 sport|chess;245;
-railway|spur;246;
+# Was railway|spur -- June 2023
+deprecated|deprecated;246;x
```
I'm not sure if its a long time or not, maybe there were some usages until recently.
So I preferred to be on the safe side.

vng replied:
Hm, we have free value 65!
Ok, put main into 65 also with comment about 1M entries.

pastk replied:
Done!

vng commented on data/styles/clear/include/Roads.mapcss:
```diff
@@ -1092,8 +1084,8 @@ line|z15-[aeroway=taxiway]
-line|z11-[railway=rail],
-line|z11-[railway=yard],
+/* railway-motor_vehicle is a non-drawable style used in routing only. */
+line|z11-[railway=rail][!motor_vehicle],
```
After mapping [!motor_vehicle] won't needed here and below.

vng commented on indexer/ftypes_matcher.cpp:
```diff
@@ -31,9 +31,10 @@ class HighwayClasses
-    /// @todo Wow, actual highway type is railway-rail-motor_vehicle.
+    /// @todo Wow, actual highway type is railway-rail-motor_vehicle (to be deprecated).
     /// Should be carefull with GetHighwayClass function.
     m_map[c.GetTypeByPath({"railway", "rail"})] = ftypes::HighwayClass::Transported;
```
Remove this entry

vng commented on indexer/ftypes_matcher.cpp:
```diff
@@ -776,6 +777,7 @@ IsWayWithDurationChecker::IsWayWithDurationChecker() : BaseChecker(3 /* level */
 {
   base::StringIL const types[] = {{"route", "ferry"},
                                   {"railway", "rail", "motor_vehicle"},
```
Remove this entry

vng commented on routing_common/vehicle_model.hpp:
```diff
@@ -55,8 +56,9 @@ enum class HighwayType : uint16_t
-  RailwayRailMotorVehicle = 994,
+  RailwayRailMotorVehicle = 994, // TODO: to be deprecated, see mapcss-mapping.csv.
   RouteShuttleTrain = 1054,
+  RailwayMotorVehicle = 1501,
```
Do not add new entry, just change value of the existing one. To avoid bloat changes.
Also don't rename.

pastk commented:
- main, service, branch and spur are in the "golden 7-bit" :) ids range now
- replaced old `railway-rail-motor_vehicle` with `railway-rail`, removed routing code specific to it
